### PR TITLE
Make some fatal lexer errors recoverable

### LINKED ISSUE #12834: rustdoc fails unpleasantly when it can't tokenise a code block
Consider a file like this:

```rust
#[crate_id = "sad"];

/// ```
/// a + b ∈ Self ∀ a, b ∈ Self
/// ```
pub fn foo() {}
```

Here's rustdoc's opinion of this:

```
<stdin>:1:7: 1:8 error: unknown start of token: \u2208
<stdin>:1 a + b ∈ Self ∀ a, b ∈ Self
                ^
task '<unnamed>' failed at '~Any', /home/chris/vc/rust/src/libsyntax/diagnostic.rs:48
```

☹

You see, it attempted to tokenise the contents of that code block.

The current solution for this is marking that code block as "notrust":

```rust
/// ```notrust
/// a + b ∈ Self ∀ a, b ∈ Self
/// ```
pub fn foo() {}
```

rustdoc should do one of two things:

1. Permit tokenisation to fail silently, discarding the notion of syntax highlighting should it fail; or

2. Retain the current notrust-must-be-explicit behaviour, but with two alterations:

    1. Give meaningful spans, if at all possible—`sad.rs:4:11: 4:12` as the error span rather than `<stdin>:1:7: 1:7`.

    2. Give you a hint that that might be want you meant.

    In total, perhaps something like this:

    ```
<stdin>:1:7: 1:8 error: unknown start of token: \u2208
<stdin>:1 a + b ∈ Self ∀ a, b ∈ Self
                    ^
sad.rs:3:5: 3:8 info: that code block does not tokenise as Rust code; perhaps you meant to mark it as notrust?
                          (e.g. use the opening fence "```notrust")
sad.rs:3 /// ```
                 ^~~
    ```

### PR DESCRIPTION
**Pull request by mitaa**
I've kept the changes to a minimum since I'm not really sure if this approach is a acceptable.

fixes #12834

cc @nrc

### CODE REVIEW

File: src/libsyntax/parse/lexer/mod.rs [hunk 1]
```diff
@@ -27,26 +27,47 @@ pub use ext::tt::transcribe::{TtReader, new_tt_reader, new_tt_reader_with_doc_fl
 pub mod comments;
 mod unicode_chars;
 
+pub type TokenAndSpanResult = Result<TokenAndSpan, ()>;
```
Comment (nrc) [inline src/libsyntax/parse/lexer/mod.rs]: This doesn't seem worth making a type for, it only saves a few chars.

File: src/libsyntax/parse/lexer/mod.rs [hunk 2]
```diff
@@ -27,26 +27,47 @@ pub use ext::tt::transcribe::{TtReader, new_tt_reader, new_tt_reader_with_doc_fl
...
+                    t = self.try_next_token()?;
                 }
                 _ => break,
             }
         }
-        t
+        Ok(t)
+    }
+    fn real_token(&mut self) -> TokenAndSpan {
+        match self.try_real_token() {
```
Comment (nrc) [inline src/libsyntax/parse/lexer/mod.rs]: could factor out the logic here and in next_token into its own method taking `self.try_real_token()` as a closure.

**Comment by nrc** [comment]
There are some small things to change inline, r+ with those fixed.

**Comment by mitaa** [comment]
I tried using a closure, but that didn't really work out.
These methods take `&mut self`, so the closure can't capture `self` which leads to passing `self` as argument of the closure making the trait not object safe. (mentioning `Self` in the signature or requiring that `Self: Sized` for converting to a trait object)

Passing just the result to `unwrap_or_abort` works, but the borrow checker isn't smart enough yet for `self.unwrap_or_abort(self.try_real_token())`, requiring a temporary.

I've also added a note to the rustdoc warning.

File: src/test/rustdoc/issue-12834.rs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
```
Comment (nrc) [inline src/test/rustdoc/issue-12834.rs]: Could you add comment to this test explaining what is being tested please?

**Comment by nrc** [comment]
@mitaa thanks for the changes and explanation, I have one more very minor request (sorry I missed it the first time round), then r+ for reals.

**Comment by mitaa** [comment]
(updated)

**Pull request merged**